packages/metrics/metrics.go

```
@@ -0,0 +1,61 @@
+package metrics
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/iotaledger/hive.go/logger"
+	"github.com/iotaledger/wasp/packages/iscp"
+	"github.com/labstack/echo/v4"
+	"github.com/labstack/echo/v4/middleware"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+type Metrics struct {
+	server                  *http.Server
+	log                     *logger.Logger
+	offLedgerRequestCounter *prometheus.CounterVec
+	onLedgerRequestCounter  *prometheus.CounterVec
+	processedRequestCounter *prometheus.CounterVec
+}
+
+type chainMetrics struct {
+	metrics *Metrics
+	chainID *iscp.ChainID
+}
+
+func (m *Metrics) NewChainMetrics(chainID *iscp.ChainID) ChainMetrics {
+	return &chainMetrics{
+		metrics: m,
+		chainID: chainID,
+	}
+}
+
+func New(log *logger.Logger) *Metrics {
+	return &Metrics{log: log}
+}
+
+func (m *Metrics) Start(addr string) error {
+	if m.server == nil {
+		e := echo.New()
+		e.HideBanner = true
+		e.Use(middleware.Recover())
+		e.GET("/metrics", func(c echo.Context) error {
+			handler := promhttp.Handler()
+			handler.ServeHTTP(c.Response(), c.Request())
+			return nil
+		})
+		m.server = &http.Server{Addr: addr, Handler: e}
+	}
+	m.registerMempoolMetrics()
+	m.log.Infof("Prometheus metrics accessible at: %s", addr)
+	return m.server.ListenAndServe()
```

could use `sync.once.Do`. Even with the nil check there is a possibility of a race condition if Start called multiple times

---

About this, `once.Do` takes an argument of type `func()` meanwhile `ListenAndServe` has type `func() error`. So to get this to work I had to wrap `ListenAndServe` inside a function and then pass that function to `once.Do`. But that hides the error.